library/core/src/option.rs

```diff
@@ -891,6 +891,16 @@ impl<T> Option<T> {
     #[rustc_const_unstable(feature = "const_option_ext", issue = "91930")]
     pub const unsafe fn unwrap_unchecked(self) -> T {
         debug_assert!(self.is_some());
+        // Make things easier for the optimizer when niches are involved
+        if mem::size_of::<T>() == mem::size_of::<Self>() {
+            // SAFETY: Size equality implies niches are involved. And with niches
+            // transmutes are ok because they don't change bits, only make use of invalid values
+            unsafe {
+                let val = mem::transmute_copy(&self);
```

Comment:
YMMV: if you want to save the separate `forget` call, I think you can write this as
```suggestion
                return mem::transmute_copy(&ManuallyDrop::new(self));
```
(Since `forget` is just putting it in a `ManuallyDrop` and ignoring it these days anyway.)